core/src/sumeragi/main_loop.rs

```diff
@@ -1266,32 +1264,48 @@ fn handle_categorized_block_sync<'state, F: Fn(PipelineEventBox)>(
     genesis_account: &AccountId,
     handle_events: &F,
     block_sync_type: Result<BlockSyncType, BlockSyncError>,
+    voting_block: &mut Option<VotingBlock>,
 ) -> Result<BlockSyncOk<'state>, (SignedBlock, BlockSyncError)> {
-    let (mut state_block, soft_fork) = match block_sync_type {
-        Ok(BlockSyncType::CommitBlock) => (state.block(), false),
-        Ok(BlockSyncType::ReplaceTopBlock) => (state.block_and_revert(), true),
+    let soft_fork = match block_sync_type {
+        Ok(BlockSyncType::CommitBlock) => false,
+        Ok(BlockSyncType::ReplaceTopBlock) => true,
         Err(e) => return Err((block, e)),
     };
-    let latest_block = state_block
-        .latest_block()
-        .expect("INTERNAL BUG: No latest block");
-    let mut topology = Topology::new(latest_block.commit_topology().cloned());
-    topology.block_committed(&latest_block, state_block.world.peers().cloned());
-    topology.nth_rotation(block.header().view_change_index as usize);
-
-    ValidBlock::validate(
+
+    let topology = {
+        let view = state.view();
+        let latest_block = if soft_fork {
+            view.prev_block().expect("INTERNAL BUG: No prev block")
+        } else {
+            view.latest_block().expect("INTERNAL BUG: No latest block")
+        };
+        let mut topology = Topology::new(latest_block.commit_topology().cloned());
+        topology.block_committed(&latest_block, view.world.peers().cloned());
+        topology.nth_rotation(block.header().view_change_index as usize);
+        topology
+    };
+
+    // Check if block have enough signatures and proxy tail signature before potentially discard previous voting block
+    if let Err(err) = ValidBlock::is_commit(&block, &topology) {
+        return Err((block, BlockSyncError::BlockNotValid(err)));
+    }
+
+    ValidBlock::validate_keep_voting_block(
```

Comment:
```suggestion
    ValidBlock::commit_keep_voting_block(
```

how about we do it like this and produce `CommitBlock` directly. Because you have a redundant check afterwards `block.commit`. Or this way:

```suggestion
    CommitedBlock::commit_keep_voting_block(
```

Reply:
I've decided to make this method of `ValidBlock` to keep all related logic nearby